README
Timeline::SVG

Create timelines using SVG.


Changes.md
# Change Log

## [0.1.2] - 2022-10-31

### Fixed

- Add DateTime::Format::Strptime as pre-req

## [0.1.1] - 2022-10-31

### Fixed

- Switch from Time::Piece to DateTime because it handles
  older dates.

## [0.1.0] - 2022-10-30

### Improved

- Support full dates (YYYY-MM-DD) instead of just years for
  start and end events (thanks simbabque and tcheukueppo).

## [0.0.10] - 2022-06-14

### Fixed

- Fixes to the way that we interact with the underlying
  SVG object (thanks choroba)

## [0.0.9] - 2020-12-22

### Fixed

- Added Time::Piece to pre-reqs

## [0.0.8] - 2020-12-21

### Fixed

- Better testing
- More subclassable

### Added

- Bugtracker info

## [0.0.7] - 2017-09-06

### Fixed

- Fixed packaging errors

## [0.0.6] - 2017-08-19

### Improved

- Generally made the class easier to subclass.

## [0.0.5] - 2017-08-19

### Improved

- Let the events draw themselves.
- Moved pod tests from t to xt.

## [0.0.4] - 2017-08-02

### Fixed

- Reverted previous shebang change as the Perl toolchain is cleverer than I thought.

## [0.0.3] - 2017-07-31

### Fixed

- Changed the command line program shebang to use `/usr/bin/env perl`

## [0.0.2] - 2017-07-31

### Fixed

- Various packaing fixes for better kwalitee

## [0.0.1] - 2017-07-29

### Added

- All the things. Release early, release often.


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME => 'SVG::Timeline',
  VERSION_FROM => 'lib/SVG/Timeline.pm',
  LICENSE      => 'perl_5',
  MIN_PERL_VERSION => '5.10.0',

  BUILD_REQUIRES => {
    'Test::More' => 0,
    'ExtUtils::MakeMaker' => 0,
    SVG => 0,
    Moose => 0,
    'List::Util' => 0,
    Carp => 0,
  },
  PREREQ_PM => {
    SVG => 0,
    Moose => 0,
    'List::Util' => 0,
    Carp => 0,
    'DateTime::Format::Strptime' => 0,
  },
  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      repository => {
        type => 'git',
        url => 'git://github.com/davorg/svg-timeline.git',
        web => 'https://github.com/davorg/svg-timeline',
      },
      bugtracker => {
        web => 'https://github.com/davorg/svg-timeline/issues',
      },
    },
  },
  EXE_FILES => [ 'bin/timeline' ],
);


lib/Timeline/SVG.pm
=head1 NAME

Timelime::SVG - Create SVG timeline charts

=head1 SYNOPSIS

    use Timeline::SVG;

    my $tl = Timeline::SVG->new;

    $tl->add_event({
      start => 1914,
      end   => 1918,
      text  => 'World War I',
    });

    $tl->add_event({
      start => 1939,
      end   => 1945,
      text  => 'World War II',
    });

    print $tl->draw;

=head1 DESCRIPTION

TODO

=head1 METHODS

=head2 new(\%options)

Creates and returns a new Timeline::SVG object. 

Takes an optional hash reference containing configuration options. You
probably don't need any of these, but the following options are supported:

=over 4

=cut


package Timeline::SVG;

use 5.010;

use Moose;
use SVG;
use List::Util qw[min max];
use Carp;

=item * events - a reference to an array containing events. Events are hash
references. See L<add_event> below for the format of events.

=cut

has events => (
  traits  => ['Array'],
  isa     => 'ArrayRef[HashRef]',
  is      => 'rw',
  default => sub { [] },
  handles => {
    all_events   => 'elements',
    add_event    => 'push',
    count_events => 'count',
    has_events   => 'count',
  },
);

=item * width - the width of the output in any format used by SVG. The default
is 100%.

=cut

has width => (
  is      => 'ro',
  isa     => 'Str',
  default => '100%',
);

=item * height - the height of the output in any format used by SVG. The
default is 100%.

=cut

has height => (
  is      => 'ro',
  isa     => 'Str',
  default => '100%',
);

=item * viewport - a viewport definition (which is a space separated list of
four integers. Unless you know what you're doing, it's probably best to leave
the class to work this out for you.

=cut

has viewbox => (
  is         => 'ro',
  isa        => 'Str',
  lazy_build => 1,
);

sub _build_viewbox {
  my $self = shift;
  return join ' ',
    $self->min_year,
    0,
    $self->years,
    ($self->bar_height * $self->count_events) + $self->bar_height
    + (($self->count_events - 1) * $self->bar_height * $self->bar_spacing);
}

=item * svg - an instance of the SVG class that is used to generate the final
SVG output. Unless you're using a subclass of this class for some reason,
there is no reason to set this manually.

=cut

has svg => (
  is         => 'ro',
  isa        => 'SVG',
  lazy_build => 1,
  handles    => [qw[xmlify line text rect cdata]],
);

sub _build_svg {
  my $self = shift;

  $_->{end} //= (localtime)[5] + 1900 foreach $self->all_events;

  return SVG->new(
    width   => $self->width,
    height  => $self->height,
    viewBox => $self->viewbox,
  );
}

=item * default_colour - the colour that is used to fill the timeline
blocks. This should be defined in the RGB format used by SVG. For example,
red would be 'RGB(255,0,0)'.

=cut

has default_colour => (
  is         => 'ro',
  isa        => 'Str',
  lazy_build => 1,
);

sub _build_default_colour {
  return 'rgb(255,127,127)';
}

=item * years_per_grid - the number of years between vertical grid lines
in the output. The default of 10 should be fine unless your timeline covers
a really long timespan.

=cut

# The number of years between vertical grid lines
has years_per_grid => (
  is      => 'ro',
  isa     => 'Int',
  default => 10, # One decade by default
);

=item * bar_height - the height of an individual timeline bar.

=cut

has bar_height => (
  is      => 'ro',
  isa     => 'Int',
  default => 50,
);

=item * bar_spacing - the height if the vertical space between bars (expresssed
as a decimal fraction of the bar height).

=cut

# Size of the vertical gap between bars (as a fraction of a bar)
has bar_spacing => (
  is      => 'ro',
  isa     => 'Num',
  default => 0.25,
);

=item * decade_line_colour - the colour of the grid lines.

=cut

# The colour that the decade lines are drawn on the chart
has decade_line_colour => (
  is      => 'ro',
  isa     => 'Str',
  default => 'rgb(127,127,127)',
);

=item * bar_outline_colour - the colour that is used for the outline of the
timeline bars.

=cut

# The colour that the bars are outlined
has bar_outline_colour => (
  is      => 'ro',
  isa     => 'Str',
  default => 'rgb(0,0,0)',
);

=back

=head2 draw_grid

=cut

sub draw_grid{
  my $self = shift;

  my $curr_year = $self->min_year;

  # Draw the grid lines
  while ( $curr_year <= $self->max_year ) {
    unless ( $curr_year % $self->years_per_grid ) {
      $self->line(
        x1           => $curr_year,
        y1           => 0,
        x2           => $curr_year,
        y2           => ($self->bar_height * ($self->count_events + 1))
                      + ($self->bar_height * $self->bar_spacing
                         * ($self->count_events - 1)),
        stroke       => $self->decade_line_colour,
        stroke_width => 1
      );
      $self->text(
        x           => $curr_year + 1,
        y           => 20,
        'font-size' => $self->bar_height / 2
      )->cdata($curr_year);
    }
    $curr_year++;
  }

  $self->rect(
     x             => $self->min_year,
     y             => 0,
     width         => $self->years,
     height        => ($self->bar_height * ($self->count_events + 1))
                    + ($self->bar_height * $self->bar_spacing
                       * ($self->count_events - 1)),
     stroke        => $self->bar_outline_colour,
    'stroke-width' => 1,
    fill           => 'none',
  );

  return $self;
}

=head2 draw

=cut

sub draw {
  my $self = shift;
  my %args = @_;

  croak "Can't draw a timeline with no events"
    unless $self->has_events;

  $self->draw_grid;

  my $curr_event_idx = 1;
  foreach ($self->all_events) {
    my $x = $_->{start};
    my $y = ($self->bar_height * $curr_event_idx)
          + ($self->bar_height * $self->bar_spacing
             * ($curr_event_idx - 1));

    $self->rect(
      x              => $x,
      y              => $y,
      width          => $_->{end} - $_->{start},
      height         => $self->bar_height,
      fill           => $_->{colour} // $self->default_colour,
      stroke         => $self->bar_outline_colour,
      'stroke-width' => 1
    );

    $self->text(
      x => $x + $self->bar_height * 0.2,
      y => $y + $self->bar_height * 0.8,
      'font-size' => $self->bar_height * 0.8,
    )->cdata($_->{text});

    $curr_event_idx++;
  }

  return $self->xmlify;
}

=head2 min_year

=cut

sub min_year {
  my $self = shift;
  return unless $self->has_events;
  my @years = map { $_->{start} } $self->all_events;
  return min(@years);
}

=head2 max_year

=cut

sub max_year {
  my $self = shift;
  return unless $self->has_events;
  my @years = map { $_->{end} // (localtime)[5] } $self->all_events;
  return max(@years);
}

=head2 years

=cut

sub years {
  my $self = shift;
  return $self->max_year - $self->min_year;
}

=head1 AUTHOR
 
Dave Cross <dave@perlhacks.com>
 
=head1 COPYRIGHT AND LICENCE
 
Copyright (c) 2017, Magnum Solutions Ltd. All Rights Reserved.
 
This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
 
=cut

1;


lib/SVG/Timeline/Event.pm
=head1 NAME

SVG::Timelime::Event - A single event in an SVG timeline.

=head1 SYNOPSIS

See L<SVG::Timeline>.

=cut

package SVG::Timeline::Event;

use 5.014;

use Moose;
use Moose::Util::TypeConstraints;
use DateTime;
use DateTime::Format::Strptime;

coerce __PACKAGE__,
  from 'HashRef',
  via  { __PACKAGE__->new($_) };

# Chosen format: yyyy-mm-dd
subtype 'SVG::Timeline::DateStr',
  as 'Str',
  where   { m/ \d{4}-\d{2}-\d{2} /ax };

subtype 'SVG::Timeline::YearMonthStr',
  as 'Str',
  where   { m/ \d{4}-\d{2} /ax };

subtype 'SVG::Timeline::YearStr',
  as 'Str',
  where   { m/ \d{4} /as };

subtype 'SVG::Timeline::Num',
  as 'Num';

coerce 'SVG::Timeline::DateStr',
  from 'SVG::Timeline::YearStr',
  via  { $_ . '-01-01' };

coerce 'SVG::Timeline::DateStr',
  from 'SVG::Timeline::YearMonthStr',
  via  { $_ . '-01' };

coerce 'SVG::Timeline::Num',
  from 'SVG::Timeline::DateStr',
  via  \&str2num;

has index => (
  is => 'ro',
  isa => 'Int',
  required => 1,
);

has text => (
  is => 'ro',
  isa => 'Str',
  required => 1,
);

sub str2num {
  my ($datestr) = @_;

  my $date = DateTime::Format::Strptime->new( pattern => '%Y-%m-%d' )
               ->parse_datetime($datestr);

  return $date->year + ( $date->day_of_year / ($date->is_leap_year ? 366 : 365) );
}

has start => (
  is => 'ro',
  isa => 'SVG::Timeline::Num',
  required => 1,
  coerce => 1,
);

has end => (
  is => 'ro',
  isa => 'SVG::Timeline::Num',
  required => 1,
  coerce => 1,
);

has colour => (
  is => 'ro',
  isa => 'Maybe[Str]',
  required => 0,
);

=head1 METHODS

=head2 start_year

Return the year that the event started.

=cut

sub start_year {
  my $self = shift;

  return int $self->start;
}

=head2 end_year

Return the year that the event ended.

=cut

sub end_year {
  my $self = shift;

  if ($self->end) {
    return int $self->end;
  } else {
    return DateTime->now->year;
  }
}

=head2 draw_on($tl)

Draw the event inside the given timeline object.

=cut

sub draw_on {
  my $self = shift;
  my ($tl) = @_;

  my $x = $self->start * $tl->units_per_year;
  my $y = ($tl->bar_height * $self->index)
        + ($tl->bar_height * $tl->bar_spacing
           * ($self->index - 1));

  $tl->rect(
    x              => $x,
    y              => $y,
    width          => ($self->end - $self->start) * $tl->units_per_year,
    height         => $tl->bar_height,
    fill           => $self->colour // $tl->default_colour,
    stroke         => $tl->bar_outline_colour,
    'stroke-width' => 1
  );

  $tl->text(
    x => ($x + $tl->bar_height * 0.2),
    y => $y + $tl->bar_height * 0.8,
    'font-size' => $tl->bar_height * 0.8,
  )->cdata($self->text);
}

no Moose;
__PACKAGE__->meta->make_immutable;

=head1 AUTHOR

Dave Cross <dave@perlhacks.com>

=head1 COPYRIGHT AND LICENCE

Copyright (c) 2017, Magnum Solutions Ltd. All Rights Reserved.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/SVG/Timeline.pm
=head1 NAME

SVG::Timeline - Create SVG timeline charts

=head1 SYNOPSIS

    use SVG::Timeline;

    my $tl = SVG::Timeline->new;

    $tl->add_event({
      start => 1914,
      end   => 1918,
      text  => 'World War I',
    });

    $tl->add_event({
      start => '1939-09-01', # 1 Sep 1939
      end   => '1945-05-08', # 8 May 1945
      text  => 'World War II',
    });

    print $tl->draw;

=head1 DESCRIPTION

This module allows you to easily create SVG documents that represent timelines.

An SVG timeline is a diagram with a list of years across the top and a
number of bars below. Each bar represents a period of time.

=head1 METHODS

=head2 new(\%options)

Creates and returns a new SVG::Timeline object.

Takes an optional hash reference containing configuration options. You
probably don't need any of these, but the following options are supported:

=over 4

=cut

package SVG::Timeline;

use 5.014;

our $VERSION = '0.1.3';

use Moose;
use Moose::Util::TypeConstraints;
use SVG;
use List::Util qw[min max];
use Carp;

use SVG::Timeline::Event;

subtype 'ArrayOfEvents', as 'ArrayRef[SVG::Timeline::Event]';

coerce 'ArrayOfEvents',
  from 'HashRef',
  via { [ SVG::Timeline::Event->new($_) ] },
  from 'ArrayRef[HashRef]',
  via { [ map { SVG::Timeline::Event->new($_) } @$_ ] };

=item * events - a reference to an array containing events. Events are hash
references. See L<add_event> below for the format of events.

=cut

has events => (
  traits  => ['Array'],
  isa     => 'ArrayOfEvents',
  is      => 'rw',
  coerce  => 1,
  default => sub { [] },
  handles => {
    all_events   => 'elements',
    add_event    => 'push',
    count_events => 'count',
    has_events   => 'count',
  },
);

around 'add_event' => sub {
  my $orig = shift;
  my $self = shift;

  $self->_clear_viewbox;
  $self->_clear_svg;

  my $index = $self->count_events + 1;
  $_[0]->{index} = $index;

  $self->$orig(@_);
};

=item * width - the width of the output in any format used by SVG. The default
is 100%.

=cut

has width => (
  is      => 'ro',
  isa     => 'Str',
  default => '100%',
);

=item * height - the height of the output in any format used by SVG. The
default is 100%.

=cut

has height => (
  is      => 'ro',
  isa     => 'Str',
  default => '100%',
);

=item * aspect_ratio - the default is 16/9.

=cut

has aspect_ratio => (
  is => 'ro',
  isa => 'Num',
  default => 16/9,
);

=item * viewport - a viewport definition (which is a space separated list of
four integers. Unless you know what you're doing, it's probably best to leave
the class to work this out for you.

=cut

has viewbox => (
  is         => 'ro',
  isa        => 'Str',
  lazy_build => 1,
  clearer    => '_clear_viewbox',
);

sub _build_viewbox {
  my $self = shift;
  return join ' ',
    $self->min_year * $self->units_per_year,
    0,
    $self->years * $self->units_per_year,
    ($self->bar_height * $self->events_in_timeline) + $self->bar_height
    + (($self->count_events - 1) * $self->bar_height * $self->bar_spacing);
}

=item * svg - an instance of the SVG class that is used to generate the final
SVG output. Unless you're using a subclass of this class for some reason,
there is no reason to set this manually.

=cut

has svg => (
  is         => 'ro',
  isa        => 'SVG',
  lazy_build => 1,
  clearer    => '_clear_svg',
  handles    => [qw[xmlify line text rect cdata]],
);

sub _build_svg {
  my $self = shift;

  $_->{end} //= (localtime)[5] + 1900 foreach $self->all_events;

  return SVG->new(
    width   => $self->width,
    height  => $self->height,
    viewBox => $self->viewbox,
  );
}

=item * default_colour - the colour that is used to fill the timeline
blocks. This should be defined in the RGB format used by SVG. For example,
red would be 'RGB(255,0,0)'.

=cut

has default_colour => (
  is         => 'ro',
  isa        => 'Str',
  lazy_build => 1,
);

sub _build_default_colour {
  return 'rgb(255,127,127)';
}

=item * years_per_grid - the number of years between vertical grid lines
in the output. The default of 10 should be fine unless your timeline covers
a really long timespan.

=cut

# The number of years between vertical grid lines
has years_per_grid => (
  is      => 'ro',
  isa     => 'Int',
  default => 10, # One decade by default
);

=item * bar_height - the height of an individual timeline bar.

=cut

has bar_height => (
  is      => 'ro',
  isa     => 'Int',
  default => 50,
);

=item * bar_spacing - the height if the vertical space between bars (expresssed
as a decimal fraction of the bar height).

=cut

has bar_spacing => (
  is      => 'ro',
  isa     => 'Num',
  default => 0.25,
);

=item * decade_line_colour - the colour of the grid lines.

=cut

has decade_line_colour => (
  is      => 'ro',
  isa     => 'Str',
  default => 'rgb(127,127,127)',
);

=item * bar_outline_colour - the colour that is used for the outline of the
timeline bars.

=cut

has bar_outline_colour => (
  is      => 'ro',
  isa     => 'Str',
  default => 'rgb(0,0,0)',
);

=back

=head2 events_in_timeline

The number of events that we need to make space for in the timeline. This
is generally just the number of events that we have added to the timeline, but
this method is here in case subclasses want t odo something different.

=cut

sub events_in_timeline {
  return $_[0]->count_events;
}

=head2 add_event

Takes a hash reference with event details and adds an L<SVG::Timeline::Event>
to the timeline. The following details are supported:

=over 4

=item * text - the name of the event that is displayed on the bar. This is required.

=item * start - the start year of the event. It is a string of format C<YYYY-MM-DD>.
For example, C<2017-07-02> is the 2nd of July 2017. The month and day can be omitted
(in which case they are replaced with '01').

=item * end - the end year of the event, requirements are the same as that of C<start>.
 
=item * colour - the colour that is used to fill the timeline block. This should be
defined in the RGB format used by SVG. For example, red would be 'RGB(255,0,0)'.
This is optional. If not provided, the C<default_color> is used.

=back

=head2 calculated_height

The height of the timeline in "calculated units".

=cut

sub calculated_height {
  my $self = shift;

  # Number of events ...
  my $calulated_height = $self->events_in_timeline;
  # ... plus one for the header ...
  $calulated_height++;
  # ... multiplied by the bar height...
  $calulated_height *= $self->bar_height;
  # .. add spacing.
  $calulated_height += $self->bar_height * $self->bar_spacing *
                       ($self->events_in_timeline - 1);

  return $calulated_height;
}

=head2 calculated_width

The width in "calulated units".

=cut

sub calculated_width {
  my $self = shift;

  return $self->calculated_height * $self->aspect_ratio;
}

=head2 units_per_year

The number of horizontal units that each year should take up.

=cut

sub units_per_year {
  my $self = shift;

  return $self->calculated_width / $self->years;
}

=head2 draw_grid

Method to draw the underlying grid.

=cut

sub draw_grid{
  my $self = shift;

  my $curr_year = $self->min_year;
  my $units_per_year = $self->units_per_year;

  # Draw the grid lines
  while ( $curr_year <= $self->max_year ) {
    unless ( $curr_year % $self->years_per_grid ) {
      $self->line(
        x1           => $curr_year * $units_per_year,
        y1           => 0,
        x2           => $curr_year * $units_per_year,
        y2           => $self->calculated_height,
        stroke       => $self->decade_line_colour,
        stroke_width => 1
      );
      $self->text(
        x           => ($curr_year + 1) * $units_per_year,
        y           => 20,
        'font-size' => $self->bar_height / 2
      )->cdata($curr_year);
    }
    $curr_year++;
  }

  $self->rect(
     x             => $self->min_year * $units_per_year,
     y             => 0,
     width         => $self->years * $units_per_year,
     height        => ($self->bar_height * ($self->events_in_timeline + 1))
                    + ($self->bar_height * $self->bar_spacing
                       * ($self->events_in_timeline - 1)),
     stroke        => $self->bar_outline_colour,
    'stroke-width' => 1,
    fill           => 'none',
  );

  return $self;
}

=head2 draw

Method to draw the timeline.

=cut

sub draw {
  my $self = shift;
  my %args = @_;

  croak "Can't draw a timeline with no events"
    unless $self->has_events;

  $self->draw_grid;

  my $curr_event_idx = 1;
  foreach ($self->all_events) {

    $_->draw_on($self);

    $curr_event_idx++;
  }

  return $self->xmlify;
}

=head2 min_year

Returns the minimum year from all the events in the timeline.

=cut

sub min_year {
  my $self = shift;
  return unless $self->has_events;
  my @years = map { $_->start_year } $self->all_events;
  return min(@years);
}

=head2 max_year

Returns the maximum year from all the events in the timeline.

=cut

sub max_year {
  my $self = shift;
  return unless $self->has_events;
  my @years = map { $_->end_year } $self->all_events;
  return max(@years);
}

=head2 years

The number of years that all the events in the timeline span.

=cut

sub years {
  my $self = shift;
  return $self->max_year - $self->min_year;
}

no Moose;
__PACKAGE__->meta->make_immutable;

=head1 AUTHOR

Dave Cross <dave@perlhacks.com>

=head1 COPYRIGHT AND LICENCE

Copyright (c) 2017, Magnum Solutions Ltd. All Rights Reserved.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
